### PR TITLE
Use provider's DNS for consumer connection

### PR DESCRIPTION
**Pull request by soffokl**

### CODE REVIEW

File: services/openvpn/service/factory.go [hunk 1]
```diff
@@ -135,6 +136,10 @@ func newSessionConfigNegotiatorFactory(networkOptions node.OptionsNetwork, servi
 		return &OpenvpnConfigNegotiator{
 			natEventGetter: natEventGetter,
 			vpnConfig: &openvpn_service.VPNConfig{
+				DNS: providerIP(net.IPNet{
```
Comment (Waldz) [inline services/openvpn/service/factory.go]: Should be just DNS endpoint `DNS: "10.0.0.1:53"` or `DNS: ["10.0.0.1:53", "10.0.0.2:53"]`
What's the point to communicate virtual network information?

Comment (soffokl) [reply services/openvpn/service/factory.go]: It's just an IP `10.8.0.1`, nothing more.

File: services/openvpn/client_config.go [hunk 2]
```diff
@@ -71,23 +72,26 @@ func defaultClientConfig(runtimeDir string, scriptSearchPath string) *ClientConf
...
 	// TODO Rename `vpnConfig` to `sessionConfig`
 	err := NewDefaultValidator().IsValid(vpnConfig)
 	if err != nil {
 		return nil, err
 	}
 
-	clientFileConfig := newClientConfig(runtimeDir, configDir)
+	clientFileConfig := newClientConfig(runtimeDir, configDir, enableDNS)
+	if enableDNS && len(vpnConfig.DNS) > 0 {
+		for _, dns := range strings.Split(vpnConfig.DNS, ",") {
```
Comment (Waldz) [inline services/openvpn/client_config.go]: Just a single address, if we want list we'll extend structure to have array

File: services/openvpn/service/factory.go (hunk 1 above)
Comment (Waldz) [reply services/openvpn/service/factory.go]: Oh, You're calculating it here. Lets just inject it instead of duplicated calculation

File: cmd/flags_network.go [hunk 3]
```diff
@@ -85,6 +85,11 @@ var (
 		Name:  "experiment-natpunching",
 		Usage: "Enables experimental NAT hole punching",
 	})
+
+	enableDNSFlag = altsrc.NewBoolTFlag(cli.BoolTFlag{
```
Comment (Waldz) [inline cmd/flags_network.go]: As it's connection choice, the best would be to have it as option passed to connection creation, instead global flag

File: services/openvpn/service/factory.go (hunk 1 above)
Comment (soffokl) [reply services/openvpn/service/factory.go]: Changed

File: services/openvpn/client_config.go (hunk 2 above)
Comment (soffokl) [reply services/openvpn/client_config.go]: changed

File: cmd/flags_network.go (hunk 3 above)
Comment (soffokl) [reply cmd/flags_network.go]: Moved to connection options.

File: services/openvpn/service/manager.go [hunk 4]
```diff
@@ -50,7 +49,7 @@ type ServerFactory func(*openvpn_service.ServerConfig, chan openvpn.State) openv
 type ProposalFactory func(currentLocation market.Location) market.ServiceProposal
 
 // SessionConfigNegotiatorFactory initiates ConfigProvider instance during runtime
-type SessionConfigNegotiatorFactory func(secPrimitives *tls.Primitives, outboundIP, publicIP string, port int) session.ConfigNegotiator
+type SessionConfigNegotiatorFactory func(secPrimitives *tls.Primitives, serverIP, outboundIP, publicIP string, port int) session.ConfigNegotiator
```
Comment (Waldz) [inline services/openvpn/service/manager.go]: `serverIP`  -> `dnsIP` probably

File: core/node/options.go [hunk 5]
```diff
@@ -47,8 +47,9 @@ type Options struct {
 	Location   OptionsLocation
 	Transactor OptionsTransactor
 
-	Openvpn  Openvpn
-	Firewall OptionsFirewall
+	EnableDNS bool
```
Comment (Waldz) [inline core/node/options.go]: Not needed anymore

File: core/node/options_network.go [hunk 6]
```diff
@@ -32,5 +32,7 @@ type OptionsNetwork struct {
 	EtherClientRPC       string
 	EtherPaymentsAddress string
 
+	EnableDNS bool
```
Comment (Waldz) [inline core/node/options_network.go]: Not needed anymore

File: services/openvpn/service/manager.go (hunk 4 above)
Comment (soffokl) [reply services/openvpn/service/manager.go]: Right, changed it.

File: core/node/options.go (hunk 5 above)
Comment (soffokl) [reply core/node/options.go]: Thanks, remove it.

File: core/node/options_network.go (hunk 6 above)
Comment (soffokl) [reply core/node/options_network.go]: Thanks, remove it.

File: services/wireguard/service/service_unix.go [hunk 7]
```diff
@@ -89,6 +91,19 @@ func (manager *Manager) ProvideConfig(sessionConfig json.RawMessage, traversalPa
 		return nil, err
 	}
 
+	config.Consumer.DNS = providerIP(config.Consumer.IPAddress).String()
+	dnsServer := dns.NewServer(
+		net.JoinHostPort(config.Consumer.DNS, "53"),
+		dns.ResolveViaConfigured(),
+	)
+
+	go func() {
```
Comment (Waldz) [inline services/wireguard/service/service_unix.go]: I dont think it's a good idea to not to care if DNS wasn't started

File: services/wireguard/service/service_unix.go [hunk 8]
```diff
@@ -100,11 +115,14 @@ func (manager *Manager) ProvideConfig(sessionConfig json.RawMessage, traversalPa
 	}
 
 	destroy := func() {
+		if err := dnsServer.Stop(); err != nil {
+			_ = log.Error("failed to stop DNS server", err)
+		}
 		if err := manager.natService.Del(natRule); err != nil {
-			log.Error("failed to delete NAT forwarding rule: ", err)
+			_ = log.Error("failed to delete NAT forwarding rule: ", err)
```
Comment (Waldz) [inline services/wireguard/service/service_unix.go]: Is this change really needed here? If not using variable, lets not define it as `_` at all

File: services/wireguard/service/service_unix.go [hunk 9]
```diff
@@ -127,3 +145,12 @@ func (manager *Manager) Stop() error {
 	log.Info("wireguard service stopped")
 	return nil
 }
+
+func providerIP(subnet net.IPNet) net.IP {
```
Comment (Waldz) [inline services/wireguard/service/service_unix.go]: Looks very complicated function, I suppose we already had one in this package

File: services/wireguard/service/service_unix.go (hunk 7 above)
Comment (soffokl) [reply services/wireguard/service/service_unix.go]: `dnsServer.Run()` is a blocking function, it returns nothing until fails. We cannot wait it forever, added logging to be able to track any problem with DNS.

File: services/wireguard/service/service_unix.go (hunk 8 above)
Comment (soffokl) [reply services/wireguard/service/service_unix.go]: `log.Error` return and error and linter complain about it. Explicitly marking that we don't care about this error.

File: services/wireguard/service/service_unix.go (hunk 9 above)
Comment (soffokl) [reply services/wireguard/service/service_unix.go]: We have similar functions in a number of places, but all of them a bit different. We probably need some general and complex function that will work for all cases. I didn't want to add it right now.

Comment (soffokl) [reply services/wireguard/service/service_unix.go]: Added this function to the utils package as a single place.

File: services/wireguard/service/service_unix.go (hunk 7 above)
Comment (Waldz) [reply services/wireguard/service/service_unix.go]: Ups, then need to do the same in Openvpn DNS server

File: services/wireguard/service/service_unix.go (hunk 8 above)
Comment (Waldz) [reply services/wireguard/service/service_unix.go]: I would say, lets change our abstraction `log.Logger` instead

File: services/wireguard/service/service_unix.go (hunk 7 above)
Comment (soffokl) [reply services/wireguard/service/service_unix.go]: Changed for OpenVPN too.

File: services/wireguard/service/service_unix.go (hunk 8 above)
Comment (soffokl) [reply services/wireguard/service/service_unix.go]: Changed.
